File: app/src/main/java/com/mmtran/turtlesoccer/loaders/FirestoreLoader.java
package com.mmtran.turtlesoccer.loaders;

import android.util.Log;

import androidx.annotation.NonNull;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.google.firebase.firestore.QuerySnapshot;
import com.google.firebase.firestore.core.OrderBy;
import com.mmtran.turtlesoccer.models.AssociationsViewModel;
import com.mmtran.turtlesoccer.models.Nation;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

public class FirestoreLoader {

    private static final String TAG = "FirestoreLoader";

    private final FirebaseFirestore db;
    private final List<Nation> nationList = new ArrayList<>();

    public FirestoreLoader() {
        db = FirebaseFirestore.getInstance();
    }

    public void getActiveNations(AssociationsViewModel associationsViewModel) {

        Query query = db.collection("nation")
                .whereEqualTo("parent_nation_id", "")
                .whereNotEqualTo("confederation_id", "");
        query.get().addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
            @Override
            public void onComplete(@NonNull Task<QuerySnapshot> task) {
                if (task.isSuccessful()) {
                    for (QueryDocumentSnapshot document : task.getResult()) {
                        Nation nation = document.toObject(Nation.class);
                        nation.setId(document.getId());
                        nationList.add(nation);
                        Log.d(TAG, document.getId() + " => " + document.getData());
                    }
                    Collections.sort(nationList, new Comparator<Nation>() {
                        @Override
                        public int compare(Nation lhs, Nation rhs) {
                            return Integer.compare(lhs.getName().compareTo(rhs.getName()), 0);
                        }
                    });
                    associationsViewModel.setNationList(nationList);
                } else {
                    Log.d(TAG, "Error getting documents: ", task.getException());
                }
            }
        });
    }

    public void getDocument(AssociationsViewModel associationsViewModel) {

        DocumentReference docRef = db.collection("nation").document("AFG");
        docRef.get().addOnSuccessListener(new OnSuccessListener<DocumentSnapshot>() {
            @Override
            public void onSuccess(DocumentSnapshot documentSnapshot) {
                Nation nation = documentSnapshot.toObject(Nation.class);
                nationList.add(nation);
                nationList.stream().filter(n ->
                        n.getParentNationId().isEmpty() && n.getConfederationId() != null && !n.getConfederationId().isEmpty())
                        .collect(Collectors.toList());
                associationsViewModel.setNationList(nationList);
            }
        });
    }
}

File: app/src/main/java/com/mmtran/turtlesoccer/loaders/FirestoreLoader.kt
package com.mmtran.turtlesoccer.loaders

import android.util.Log

import com.google.firebase.firestore.*
import com.mmtran.turtlesoccer.models.*

import java.util.*

class FirestoreLoader {

    private val db: FirebaseFirestore = FirebaseFirestore.getInstance()
    private val confederationList: MutableList<Confederation> = ArrayList()
    private val competitionList: MutableList<Competition> = ArrayList()
    private val tournamentList: MutableList<Tournament> = ArrayList()
    private val nationList: MutableList<Nation> = ArrayList()
    private val teamList: MutableList<Team> = ArrayList()

    fun getNations(nationListViewModel: NationListViewModel) {

        val query = db.collection("nation")
        query.get().addOnCompleteListener { task ->
            if (task.isSuccessful) {
                for (document in task.result!!) {
                    val nation = document.toObject(Nation::class.java)
                    nation.id = document.id
                    nationList.add(nation)
//                    Log.d(TAG, document.id + " => " + document.data)
                }
                nationList.sortWith { lhs, rhs ->
                    lhs!!.name.compareTo(rhs!!.name).compareTo(0)
                }
                nationListViewModel.setNationList(nationList)
            } else {
                Log.d(TAG, "Error getting documents: ", task.exception)
            }
        }
    }

    fun getTeams(teamListViewModel: TeamListViewModel) {

        val query = db.collection("team")
        query.get().addOnCompleteListener { task ->
            if (task.isSuccessful) {
                for (document in task.result!!) {
                    val team = document.toObject(Team::class.java)
                    team.id = document.id
                    teamList.add(team)
//                    Log.d(TAG, document.id + " => " + document.data)
                }
                teamList.sortWith { lhs, rhs ->
                    lhs!!.name!!.compareTo(rhs!!.name!!).compareTo(0)
                }
                teamListViewModel.setTeamList(teamList)
            } else {
                Log.d(TAG, "Error getting documents: ", task.exception)
            }
        }
    }

    fun getConfederations(confederationListViewModel: ConfederationListViewModel) {

        val query: Query = db.collection("confederation")
        query.get().addOnCompleteListener { task ->
            if (task.isSuccessful) {
                var fifa = Confederation()
                for (document in task.result!!) {
                    val confederation = document.toObject(
                        Confederation::class.java
                    )
                    confederation.id = document.id
                    if (confederation.id == "FIFA") {
                        fifa = confederation
                    } else {
                        confederationList.add(confederation)
                    }
//                    Log.d(TAG, document.id + " => " + document.data)
                }
                confederationList.add(0, fifa)
                confederationListViewModel.setConfederationList(confederationList)
            } else {
                Log.d(TAG, "Error getting documents: ", task.exception)
            }
        }
    }

    fun getCompetitions(competitionsViewModel: CompetitionListViewModel) {

        val query: Query = db.collection("competition")
            .whereNotEqualTo("order", -1)
        query.get().addOnCompleteListener { task ->
            if (task.isSuccessful) {
                for (document in task.result!!) {
                    val competition = document.toObject(
                        Competition::class.java
                    )
                    competition.id = document.id
                    competitionList.add(competition)
//                    Log.d(TAG, document.id + " => " + document.data)
                }
                competitionsViewModel.setCompetitionList(competitionList)
            } else {
                Log.d(TAG, "Error getting documents: ", task.exception)
            }
        }
    }

    fun getTournaments(tournamentListViewModel: TournamentListViewModel) {

        val query: Query = db.collection("tournament")
        query.get().addOnCompleteListener { task ->
            if (task.isSuccessful) {
                for (document in task.result!!) {
                    val tournament = document.toObject(
                        Tournament::class.java
                    )
                    tournament.id = document.id
                    tournamentList.add(tournament)
//                    Log.d(TAG, document.id + " => " + document.data)
                }
                tournamentListViewModel.setTournamentList(tournamentList)
            } else {
                Log.d(TAG, "Error getting documents: ", task.exception)
            }
        }
    }

//    fun getNation(competition: Competition) {
//
//        val docRef = db.collection("nation").document("FRA")
//        docRef.get().addOnSuccessListener { documentSnapshot ->
//            val nation = documentSnapshot.toObject(Nation::class.java)
//            nation!!.id = documentSnapshot.id
//            competition.currentChampionNation = nation
//        }
//    }

    companion object {
        private const val TAG = "FirestoreLoader"
    }
}
